File: solutions/lua/wordy/1/wordy.lua
local re = {
  first = "^(-?%d+)",
  add = "^ plus (-?%d+)",
  sub = "^ minus (-?%d+)",
  mul = "^ multiplied by (-?%d+)",
  div = "^ divided by (-?%d+)",
}
return {
  answer = function(question)
    question = question:gsub("^What is ", "")
    local result = tonumber(question:match(re.first))
    question = question:gsub(re.first, "")
    while question ~= "?" do
      local add, sub = question:match(re.add), question:match(re.sub)
      local mul, div = question:match(re.mul), question:match(re.div)
      if add then
        result = result + tonumber(add)
        question = question:gsub(re.add, "")
      elseif sub then
        result = result - tonumber(sub)
        question = question:gsub(re.sub, "")
      elseif mul then
        result = result * tonumber(mul)
        question = question:gsub(re.mul, "")
      elseif div then
        result = result / tonumber(div)
        question = question:gsub(re.div, "")
      else
        error("Invalid question")
      end
    end
    return result
  end
}

File: solutions/ruby/wordy/1/wordy.rb
# frozen_string_literal: true

# class Op
class Op
  def initialize(regexp) = @regexp = regexp
  def match(line) = line[@regexp, 1]
  def val(line) = match(line).to_i
  def next(question) = question.gsub(@regexp, '')
end

# class Add
class Add < Op
  def initialize = super(/^ plus (-?\d+)/)
  def eval(a, b) = a + b
end

# class Sub
class Sub < Op
  def initialize = super(/^ minus (-?\d+)/)
  def eval(a, b) = a - b
end

# class Mul
class Mul < Op
  def initialize = super(/^ multiplied by (-?\d+)/)
  def eval(a, b) = a * b
end

# class Div
class Div < Op
  def initialize = super(/^ divided by (-?\d+)/)
  def eval(a, b) = a / b
end

# class WordProblem
class WordProblem
  attr_reader :answer

  FIRST = /^(-?\d+)/
  OPS = [Add.new, Sub.new, Mul.new, Div.new].freeze
  def initialize(question)
    question = question.gsub(/^What is /, '')
    @answer = question[FIRST].to_i
    question = question.gsub(FIRST, '')
    while question != '?'
      op = OPS.find { _1.match(question) }
      raise ArgumentError, 'Invalid question' unless op

      @answer = op.eval(@answer, op.val(question))
      question = op.next(question)
    end
  end
end
